api/src/main/java/org/apache/iceberg/OverwriteFiles.java

```diff
@@ -74,5 +90,22 @@
    *
    * @return this for method chaining
    */
-  OverwriteFiles validateAddedFiles();
+  OverwriteFiles validateAddedFilesMatchRowFilter();
+
+  /**
+   * Enables validation of files that are added concurrently.
+   * <p>
+   * This method should be called when the table is queried to determine files to delete/append.
+   * If a concurrent operation commits a new file after the data was read and that file might
+   * contain rows matching the specified row filter, the overwrite operation will detect
+   * this during retries and fail.
+   * <p>
+   * Calling this method with a correct row filter is required to maintain serializable isolation.
+   * Otherwise, the isolation level will be snapshot isolation.
+   *
+   * @param readSnapshotId the snapshot id that was used to read the data
+   * @param rowFilter an expression on rows in the table
+   * @return this for method chaining
+   */
+  OverwriteFiles validateNoConflictingAppends(Long readSnapshotId, Expression rowFilter);
```

Comment:
Can we rename `rowFilter` to `conflictDetectionFilter`?

Why use `Long` instead of `long`? That would allow passing null. Is that intentional? If null, then use the base snapshot ID?

Reply:
Allowing `null` was intentional to handle cases when the table is empty.